File: built-in plugins dsp/mb-bass-electric.cpp
/**
 * MB Electric Bass
 * Category : instrument
 * Type     : bass
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Fingered electric bass
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_BASS_ELECTRIC_H
#define MB_BASS_ELECTRIC_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbBassElectric : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-bass-electric";
    static constexpr const char* PLUGIN_NAME    = "MB Electric Bass";
    static constexpr const char* PLUGIN_TYPE    = "bass";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float tone = 0.5f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbBassElectric() = default;
    ~MbBassElectric() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.tone = std::clamp(params.tone, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Electric Bass
        return input;
    }
};

#endif // MB_BASS_ELECTRIC_H

File: built-in plugins dsp/mb-bass-electric.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbBassElectricPlugin: PluginDefinition = {
  id: "mb-bass-electric",
  slug: "mb-bass-electric",
  name: "MB Electric Bass",
  category: "instrument",
  type: "bass",
  version: "1.0.0",
  description: "Fingered electric bass",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "triangle", detune: 0, gain: 0.6 },
    { type: "sine", detune: 0, gain: 0.4 },
  ],
  envelope: { attack: 0.005, decay: 0.4, sustain: 0.5, release: 0.3 },
  parameters: [
    {
      id: "tone",
      name: "Tone",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { tone: 0.5, volume: 0.8 },
};

export default MbBassElectricPlugin;
